File: 329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.cpp
class Solution {
private:
    int m;
    int n;
    vector<vector<int>> max_vec;
    vector<vector<int>> mat;
    vector<pair<int,int> > directions = {
        make_pair(0,1),
        make_pair(0,-1),
        make_pair(1,0),
        make_pair(-1,0)
    };
    
public:
    int dfs(int i, int j) {
        if(max_vec[i][j] != -1) return max_vec[i][j];
        
        int mx = 0;
        
        for(pair<int,int> direction : directions) {
            int row = i+direction.first;
            int col = j+direction.second;
            
            if(row<0 || col<0 || row>=m || col>=n || mat[row][col]<=mat[i][j]) continue;
            
            mx = max(mx, 1+dfs(row,col));
        }
        
        max_vec[i][j] = mx;
        
        return mx;
    }
    
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        m = matrix.size();
        n = matrix[0].size();
        mat = matrix;
        max_vec = vector<vector<int>>(m, vector<int>(n, -1));
        
        int mx = 0;
        
        for(int i=0; i<m; i++) {
            for(int j=0; j<n; j++) {
                if(max_vec[i][j] != -1) continue;
                
                mx = max(mx, dfs(i,j));
            }
        }
        
        return 1+mx;
    }
};

File: 329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.py
from functools import cache

class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        row = len(matrix)
        col = len(matrix[0])
        result = 0
        result_map = {}

        def check_bounds(x,y):
            nonlocal row,col
            if x >= 0 and y >= 0 and x < row and y < col:
                return True
            else:
                return False
        @cache
        def dfs(x,y):
            temp_result = 0
            direction = [(0,1),(1,0),(-1,0),(0,-1)]
            for i,j in direction:
                new_x = x+i
                new_y = y+j
                if check_bounds(new_x,new_y) and matrix[new_x][new_y] > matrix[x][y]:
                    temp_result = max(temp_result,dfs(new_x,new_y))
            return temp_result + 1


        for i in range(row):
            for j in range(col):
                visited = []
                result = max(result,dfs(i,j))

        return result